WhateverGreen/kern_igfx_pm.cpp

```diff
@@ -0,0 +1,135 @@
+//
+//  kern_igfx_pm.cpp
+//  WhateverGreen
+//
+//  Created by Pb on 22/06/2020.
+//  Copyright © 2020 vit9696. All rights reserved.
+//
+
+#include <Headers/kern_atomic.hpp>
+#include <Headers/kern_patcher.hpp>
+#include <Headers/kern_devinfo.hpp>
+#include <Headers/kern_cpu.hpp>
+#include <Headers/kern_disasm.hpp>
+#include <Library/LegacyIOService.h>
+#include "kern_igfx.hpp"
+
+#include <kern/clock.h>
+
+namespace {
+constexpr const char* log = "igfx_pm";
+
+// For debugging
+struct [[gnu::packed]] IGHwCsDesc {
+  char type;
+  char gap[4];
+  char *title;
+  char unk0[48];
+  char unk1[12];
+};
+
+static bool patchRCSCheck(mach_vm_address_t& start) {
+	constexpr unsigned ninsts_max {256};
+	
+	hde64s dis;
+	
+	bool found_cmp = false;
+	bool found_jmp = false;
+
+	for (size_t i = 0; i < ninsts_max; i++) {
+		auto sz = Disassembler::hdeDisasm(start, &dis);
+
+		if (dis.flags & F_ERROR) {
+			SYSLOG(log, "Error disassembling submitExecList");
+			break;
+		}
+
+		/* cmp byte ptr [rcx], 0 */
+		if (!found_cmp && dis.opcode == 0x80 && dis.modrm_reg == 7 && dis.modrm_rm == 1)
+			found_cmp = true;
+		/* jnz rel32 */
+		if (found_cmp && dis.opcode == 0x0f && dis.opcode2 == 0x85) {
+			found_jmp = true;
+			break;
+		}
+
+		start += sz;
+	}
+	
+	if (found_jmp) {
+		auto status = MachInfo::setKernelWriting(true, KernelPatcher::kernelWriteLock);
+		if (status == KERN_SUCCESS) {
+			constexpr uint8_t nop6[] {0x90, 0x90, 0x90, 0x90, 0x90, 0x90};
+			lilu_os_memcpy(reinterpret_cast<void*>(start), nop6, arrsize(nop6));
+			MachInfo::setKernelWriting(false, KernelPatcher::kernelWriteLock);
+			DBGLOG(log, "Patched submitExecList");
+			return true;
+		} else {
+			DBGLOG(log, "Failed to set kernel writing");
+			return false;
+		}
+	} else {
+		SYSLOG(log, "jnz in submitExecList not found");
+		return false;
+	}
+}
+
+constexpr uint32_t GEN6_RP_STATE_CAP = 0x140000 + 0x5998;
```

Comment:
Could you add more documentation on what this does and how the value is calculated? Ideally with links to 01.org.

Reply:
This register is not publicly documented anywhere